app/controllers/api/base_controller/renderer.rb

```diff
@@ -540,6 +546,64 @@ def render_options(resource, data = {})
         klass = collection_class(resource)
         render :json => OptionsSerializer.new(klass, data).serialize
       end
+
+      # This is a helper method used by both .determine_include_for_find and
+      # .determine_extra_cols to collect and filter virtual_attributes for the
+      # :include_for_find and :extra_cols options that are passed to Rbac. The
+      # intent is to reduce a large amount of shared code between those two
+      # shared methods by combining them into this one.
+      #
+      # The required block used by each of aforementioned methods is used to do
+      # custom filtering that pertains to each of those methods.
+      #
+      def virtual_attributes_for(klass)
+        return nil unless klass.respond_to?(:reflect_on_association)
+
+        type    = @req.subject
+        results = []
+
+        validate_attr_selection(klass).last.each do |vattr|
+          next if vattr == "href_slug"
+
+          attr_name, attr_base = split_virtual_attribute(vattr)
+          filtered_attr        = yield type, attr_name, attr_base
+          results             << filtered_attr if filtered_attr
+        end
+
+        results.empty? ? nil : results
+      end
+
+      def determine_include_for_find(klass)
+        attrs = virtual_attributes_for(klass) do |type, attr_name, attr_base|
+          next if attr_base.blank?
+          next if virtual_attribute_accessor(type, attr_name)
+          next if Rbac::Filterer::CLASSES_THAT_PARTICIPATE_IN_RBAC.include?(attr_base)
+
+          attr_base
+        end
+
+        # Handle nested relationships and convert to a hash
+        if attrs
+          attrs.each_with_object({}) do |key, include_for_find|
+            if key.include?(".")
+              nested = include_for_find
+              key.split(".").each {|k| nested = nested[k] ||= {} }
```

Comment:
So this line is a little funky, so let me explain.

`key` in this case would be something like `"hardware.host"` (for an attribute of `"hardware.host.name"`).  The code here is then splitting it into `["hardware", "host"]`, and "recursively but not recursively" setting the `nested` variable (and the key at the current level) so the output ends up being:

```ruby
{"hardware" => {"host" => {}}}
```

This should also de-dup at the same time, but haven't tested this to be honest.